### PR TITLE
Rust: AST support for variables

### PR DESCRIPTION
**Pull request by hvitved**
This PR adds (initial) AST support for variables. For now, two classes `Variable` and `VariableAccess` are exposed, and a `VariableAccess` is linked to the `Variable` that it accesses.

Commit-by-commit review is suggested.

Care must be taken when taking shadowing into account: A variable can be shadowed both in a nested scope

```rust
let x = 0;
use(x);         // x = 0
{
    let x = "";
    use(x);     // x = ""
}
use(x);         // x = 0
```

as well as in the same scope

```rust
let x = 0;
use(x);     // x = 0
let x = "";
use(x);     // x = ""
```

In order to identify which variable is being accessed, we employ a purely location-based analysis, as we want to avoid an analysis that depends on the CFG.

In almost all cases, a `Variable` is identified by the `IdentPat` node that introduces it, however in case of `or` patterns like
```rust
match either {
    Either::Left(x) | Either::Right(x) => println!(x),
}
```
`x` can be introduced by either of the two branches, so we instead identify `x` with the whole `Either::Left(x) | Either::Right(x)` pattern.

### CODE REVIEW

File: rust/ql/test/utils/internal/InlineExpectationsTestImpl.qll [hunk 1]
```diff
@@ -0,0 +1,15 @@
+private import rust as R
+private import R
+private import codeql.util.test.InlineExpectationsTest
+
+module Impl implements InlineExpectationsTestSig {
+  /**
+   * A class representing line comments in C# used by the InlineExpectations core code
```
Comment (aibaars) [inline rust/ql/test/utils/internal/InlineExpectationsTestImpl.qll:7]: `C#` ? ;-)

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+      variableAccessCandInScope(cand, inner, name, nestLevel - 1, _, _, _, _) and
+      scope = getEnclosingScope(inner) and
+      // Use the location of the inner scope as the location of the access, instead of the
+      // actual access location. This allows us to collapse multiple accesses in inner
+      // scopes to a single entity
+      scope.getLocation().hasLocationInfo(_, startline, startcolumn, endline, endcolumn)
+    )
+  }
+
+  private newtype TVariableOrAccessCand =
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:230]: This type looked a little strange to me at first sight. Perhaps you could add a comment explaining the idea of algorithm. If I understand correctly you rank all variable definitions and uses, and bind each use to the highest ranking definition that has a lower rank than the variable use.

**Review by aibaars — COMMENTED** [review]
Looks good to me.

File: rust/ql/lib/codeql/rust/elements/internal/CommentImpl.qll [hunk 3]
```diff
@@ -12,12 +11,20 @@ private import codeql.rust.elements.internal.generated.Comment
...
    * /// This is a doc comment
    * ```
    */
-  class Comment extends Generated::Comment { }
+  class Comment extends Generated::Comment {
+    /**
+     * Gets the text of this comment, excluding the comment marker.
+     */
+    string getCommentText() {
+      exists(string s | s = this.getText() | result = s.regexpCapture("///?\\s*(.*)", 1))
```
[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/CommentImpl.qll:27]: Yes, but I skipped that for now, as it is not needed for inline test expectations.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+      not p0 instanceof OrPat
+    )
+  }
+
+  /** Gets the immediately enclosing `|` pattern of `p`, if any */
+  private OrPat getEnclosingOrPat(Pat p) { result = getAPatAncestor(p) }
+
+  /** Gets the outermost enclosing `|` pattern parent of `p`, if any. */
+  private OrPat getOutermostEnclosingOrPat(IdentPat p) {
+    result = getEnclosingOrPat+(p) and
```
[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:56]: Those should be handled as well; `getAPatAncestor` uses the generic `getImmediateParent` predicate which also works for `ParenPat`.

[the remark this answers is not in the document]
Comment (aibaars) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:56]: Ok, could you add a test case, just to be sure?

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 2 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:230]: Fair; I have tried to document the class with the algorithm that is used.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 4 above)
[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:56]: Sure.

**Review by aibaars — APPROVED** [review]

**Pull request merged**